### PR TITLE
perf(losses): cut the HausdorffERLoss erosion loop's per-iteration kernels

### PR DESCRIPTION
**Pull request by dchaudhari7177**
## What does this pull request do?

This takes the first two items from #4532 in `_HausdorffERLossBase.perform_erosion` and leaves the loop structure alone:

- `erosion = (dilation - 0.5).masked_fill(dilation < 0.5, 0)` replaces `erosion[erosion < 0] = 0`. It drops the data-dependent `index_put_` and keeps the boolean index's gradient of 1 at `dilation == 0.5` exactly. (The first version used `clamp_min(0)`, but its derivative at the bound is 0 on torch >= 2.14, which changed gradients on quantized predictions. Caught in review, see below.)
- The `mask` tensor was all ones and only multiplied into the `where` condition, so it's gone. The `ones_like(_range)` fallback divisor is now the scalar `1.0`.

## Related issue or discussion

Refs #4532. The third item, a `torch.compile` / `test_dynamo` case, is left for a follow-up, and a maintainer should decide whether the compile-first workflow applies to losses. I can't measure MPS here, so the numbers below are CPU only.

## What did you check?

**Equivalence.** `HausdorffERLoss` and `HausdorffERLoss3D` from this branch against `main`'s module (loaded from `git show upstream/main:`), over float32/float64 × k ∈ {1, 3, 10} × 12 seeds × {random inputs, inputs rounded to halves, inputs rounded to quarters}. The rounding forces dilations of exactly 0.5 and zero ranges. For each case I compared the loss and the gradient with respect to `pred`, on both torches:

```text
torch 2.5.0+cpu   loss+grad bit-identical 432/432
torch 2.14.0+cpu  loss+grad bit-identical 432/432
```

The earlier `clamp_min` version measured 960/960 on torch 2.5.0 only, which is why it missed the 2.14 tie. `test_gradient_at_the_threshold_tie` now pins that tie: on a float32 fixture that reaches `dilation == 0.5`, it compares the gradient against a reference loop that uses the boolean index. With `clamp_min` put back, it passes on 2.5.0 and fails in both 2D and 3D on 2.14.0.

**Launch count and CPU time** (default k=10, 2 classes, `TorchDispatchMode` counter, 15-run medians, torch 2.5.0 CPU):

| | main | head |
|---|---:|---:|
| aten calls per forward, 1×2×64² | 410 | 348 |
| 1×2×64² | 4.08 ms | 2.96 ms |
| 8×2×256² | 233 ms | 219 ms |

The removed calls are `lt`, `index_put_`, `ones_like` and the `mask` multiply, 22 per iteration pair. The launch-bound MPS case in the issue should gain the most, but I haven't measured it.

```text
$ python -m pytest tests/losses/test_hd.py --device=cpu --dtype=float32,float64   # torch 2.5.0 and 2.14.0
23 passed, 2 skipped, 1 xfailed
$ ... cpu-float16 profile: 12 passed;  cpu-bfloat16 profile: 12 passed
$ python -m pytest tests/onnx/test_export_coverage.py -k "Hausdorff or hausdorff or hd"
1 passed
$ uvx ruff@0.16.6 check / format --check
All checks passed!
```

No changelog fragment: outputs and gradients are bit-identical, so there's nothing user-visible to record. Could a maintainer apply `no-changelog`?

## How did AI help?

I used an AI assistant to write the tie-gradient test, write the equivalence and launch-count probe and draft this description. I ran the probe, the timings and the tests above and reviewed the diff.

### CODE REVIEW

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
Review of `eaa8401c`.

The launch-count reduction is real and your measurement holds up — but the `clamp_min` substitution **changes gradients on torch >= 2.14**, so the "bit-identical" claim is only true on the torch you measured on.

### [P1] `clamp_min(0)` and `erosion[erosion < 0] = 0` disagree at the bound

`torch.clamp`/`clamp_min`'s derivative *at* the bound is version-dependent (AGENTS.md, #4229). I reproduced the op-level split directly:

```
torch 2.14.0   clamp_min(0)  on [-1, 0, 1]  grad = [0.0, 0.0, 1.0]
               relu          on [-1, 0, 1]  grad = [0.0, 0.0, 1.0]
               y[y<0]=0      on [-1, 0, 1]  grad = [0.0, 1.0, 1.0]
               masked_fill   on [-1, 0, 1]  grad = [0.0, 1.0, 1.0]
               where         on [-1, 0, 1]  grad = [0.0, 1.0, 1.0]
torch 2.5.1    clamp_min and index_put both grad = [0.0, 1.0, 1.0]
```

So `x == 0` — i.e. `dilation == 0.5` exactly — is where they differ, and it happens in ordinary use. `HausdorffERLoss` + `HausdorffERLoss3D`, 576 cases (2D/3D × float32/float64 × 4 input styles × k ∈ {1,3,10} × 12 seeds), head vs main on torch 2.14.0:

```
loss bit-identical 576/576  (max |dloss| = 0)
grad bit-identical 459/576  (max |dgrad| = 0.0858025)
3360 differing gradient elements, 1152 of them with a CHANGED SIGN
worst case: |dgrad|max / |grad|max = 0.44
```

It is quantized predictions that trip it — `pred ∈ {0, 0.5, 1.0}` fails 36/36 in 2D and 36/36 in 3D, quarter-valued 27/36 and 18/36, while `randn` and every float64 group are 0/36. Over the random sweep, `dilation - 0.5 == 0` at 132 / 161280 elements. And the same 48 cases under both torches:

```
torch 2.5.1 : grad bit-identical 48/48
torch 2.14.0: grad bit-identical  6/48
```

which is exactly why your torch 2.5.0 probe reported 960/960. CI runs both legs and 2.14.0 is the default.

**`masked_fill` keeps everything you wanted** — no `lt` + data-dependent `index_put_`, no `nonzero`, no host sync — and restores bit-identity:

```python
erosion = (dilation - 0.5).masked_fill(dilation < 0.5, 0)
```

```
vs main, same 576 cases: loss bit-identical 576/576, grad bit-identical 576/576
```

### [P2] Falsified comment, and the changelog call

`hausdorff.py:73-74` says "clamp_min is one kernel with the same values **and gradient** as zeroing the negatives through a boolean index". That must not land — it is false on the default CI torch. On the changelog: with `masked_fill` in place the change is bit-identical end to end and `no-changelog` plus your body explanation is correct. If `clamp_min` stays, gradients change on torch >= 2.14 and it needs `changelog.d/4594.fixed.md` saying so.

### [P3] No test notices this

`tests/losses/test_hd.py` is 21 passed / 1 xfailed on both heads, and `gradcheck` passes on both, because random float64 input never lands on the tie. A tie-gradient pin — one crafted input where `dilation == 0.5` — is worth adding with the fix.

### Verified good

The rest of the diff is provably clean: with only `clamp_min` swapped for `masked_fill`, the `mask` removal and `torch.where(_to_norm, _range, 1.0)` are 576/576 bit-identical in both loss and gradient, including the scalar-vs-`ones_like` dtype question in float16/bfloat16.

The data-dependent op really is gone:

```
TorchDispatchMode, k=3
  main: {_local_scalar_dense: 3, index_put_: 6}     head: {_local_scalar_dense: 3}
```

The surviving `_local_scalar_dense` is the pre-existing `target.max().item()` at `hausdorff.py:116`, which is also why `fullgraph=True` still fails identically on both (4 graph breaks, 5 graphs) — no compile regression, and leaving that for a follow-up is the right call.

Timings, torch 2.14.0 / M1, `blocked_autorange` medians:

| | main | head |
|---|---:|---:|
| aten calls / forward (k=10) | 408 | 346 |
| CPU 1×2×64² | 1.45 ms | 1.35 ms |
| CPU 8×2×256² | 126.5 ms | 117.2 ms |
| MPS 1×2×64², no_grad (3 reps) | 18.7 / 18.9 / 36.1 ms | 17.5 / 14.3 / 14.1 ms |
| MPS 8×2×256², no_grad (3 reps) | 84.8 / 97.5 / 83.2 ms | 58.3 / 49.8 / 49.3 ms |

~1.07x on CPU and ~1.2–1.7x on MPS — the `index_put_` removal is the part MPS cares about, matching the premise in #4532. ONNX export also gets smaller (13691 → 11753 bytes) and `tests/onnx -k hausdorff` passes on both. Forward is bit-identical everywhere, including cpu/mps × float32/float64/float16/bfloat16 with dtype and device preserved. `Refs #4532` is correct and no `#4532` prose exists in `kornia/`.


Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by ducha-aiki** [comment]
Addendum: my review above was written against `eaa8401c` and you pushed `c169b2c3` at 15:49, just before I posted.

`changelog.d/4594.fixed.md` is now present, so the `Check changelog fragments` leg is green and my changelog point is resolved — though note which way it resolves: a fragment is the right call **only if `clamp_min` stays**, because then gradients do change on torch >= 2.14. If you take the `masked_fill` fix, the change becomes bit-identical end to end and `no-changelog` plus a body explanation is the better answer; tell me and I will apply the label.

**The P1 stands unchanged** — `hausdorff.py` still has the `clamp_min(0)` substitution, so the gradient divergence at `dilation == 0.5` (3360 differing elements, 1152 with a flipped sign, over 576 cases on torch 2.14.0) is still live, as is the falsified "same values **and gradient**" comment at `hausdorff.py:73-74` and the missing tie-gradient pin.


Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by dchaudhari7177** [comment]
Switched to `(dilation - 0.5).masked_fill(dilation < 0.5, 0)` and replaced the false comment. Since the change is now bit-identical end to end, I removed `changelog.d/4594.fixed.md`. Could you apply `no-changelog`?

Tie pin: `test_gradient_at_the_threshold_tie` (2D and 3D) uses quantized `pred ∈ {0, 0.5, 1}` and compares `pred.grad` exactly against a reference loop that uses `erosion[erosion < 0] = 0`. It asserts the fixture actually reaches `dilation == 0.5`. It only runs at float32, because the 0.2 cross kernel never sums to exactly 0.5 in float64. With `clamp_min(0)` put back, it passes on torch 2.5.0 and fails in both 2D and 3D on 2.14.0; with `masked_fill` it passes on both.

Head vs `main` (loaded from `git show upstream/main:`), float32/float64 × k {1, 3, 10} × 12 seeds × {random, halves, quarters}, loss and `pred.grad`: **432/432 bit-identical on torch 2.5.0 and on 2.14.0**. `test_hd.py` float32/float64: 23 passed, 2 skipped, 1 xfailed on both. I updated the PR body to match.

**Review by ducha-aiki — APPROVED** [review]
Re-review of `eaba2a466941`.

The threshold is now implemented as `(dilation - 0.5).masked_fill(dilation < 0.5, 0)`, which preserves the reference derivative at the exact tie on both sides of the supported PyTorch range. The new 2-D/3-D regression test reaches `dilation == 0.5` explicitly and compares the full input gradient with an indexed-assignment reference, so restoring `clamp_min(0)` fails on torch 2.14 while this implementation passes.

Validation: `pytest tests/losses/test_hd.py --device=cpu --dtype=float32,float64` (23 passed, 2 skipped, 1 expected failure); clean real merge tree against `origin/main` `3a6a1e333a39`.

Merge gate: the current changelog check is failing because this revision removes the fragment. The requested `no-changelog` exemption still needs maintainer application, or a valid fragment must be restored; this code approval does not mark CI as green.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-5.6-sol).
